hack/generate_tests.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/python
+"""Regenerate tests for only the files that have changed."""
+
+import argparse
+import logging
+import os
+import subprocess
+
+TOP_LEVEL_EXCLUDES = ["docs"]
+
+def generate_test_name(repo_root, package_dir):
+  """Generate the name of the go file to write the test to.
+
+  Args:
+    repo_root: Root of the repository
+    package_dir: Full path to the kustomize directory.
+  """
+
+  rpath = os.path.relpath(package_dir, repo_root)
+
+  pieces = rpath.split(os.path.sep)
+  name = "-".join(pieces) + "_test.go"
+
+  return name
+
+def get_changed_dirs():
+  """Return a list of directories of changed kustomization packages."""
+  # Generate a list of the files which have changed with respect to the upstream
+  # branch
+  modified_files = subprocess.check_output(
+    ["git", "diff", "--name-only", "@{upstream}"])
```

Comment:
How would prompting the user be different from what happens now? i.e. you run it, see an error about no upstream branch; you set the upstream and then rerun it.

Wouldn't having a config file be duplicating what git is doing? i.e. git is already giving you a way to configure the upstream branch.

Reply:
To clarify here: is the aim to generate the diff from the last upstream commit in the working branch?
If this is the case, a developer could forget to generate tests and push their changes upstream. The diffs wouldn't reflect the committed changes and the required tests might not get generated.

Alternatively, should we always compare against upstream/master assuming it is almost always updated?
This way, we avoid the branch locality of 'upstream' configuration changing and we can set our local configuration to upstream/master which will be independent of the branch which we are working on.

Please correct me if I'm mistaken.

Reply:
> If this is the case, a developer could forget to generate tests and push their changes upstream. The diffs wouldn't reflect the committed changes and the required tests might not get generated.

upstream should be kubeflow/manifests users should not be able to push changes to that commit because of an error.

> Alternatively, should we always compare against upstream/master assuming it is almost always updated?

Isn't that what the code is doing? The code is using a diff whatever the user set as the upstream branch to determine what files are changed in the PR and if so generate the changed files.

My expectation is users set the upstream branch to `kubeflow/manifests`; not their personal fork of the kubeflow/manifests.

See for example
https://hackernoon.com/sync-a-fork-from-upstream-repo-in-github-c2c29c8eca3b

Reply:
Got it. Thanks.